# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: widget/gtk/nsWindow.cpp
# Commit: b2f81bedfe4d
# Full Hash: b2f81bedfe4de95abca75a294b5ccea26b486ce3
# Author: stransky <stransky@redhat.com>
# Date: 2021-06-15 21:45:02
# Description:
#   Bug 1715292 [Wayland] Update popup visibility and hierarchy handling, r=jhorak
#   
#   - Don't remove popup from widget hierarchy on nsWindow::Destroy() but leave nsWindow::Show(false) do the job.
#   - Don't check popup widget visibility from Gtk as it's not reliable.
#   
# ==============================================================================

diff -r d6f7f2c7a16c -r b2f81bedfe4d widget/gtk/nsWindow.cpp
--- a/widget/gtk/nsWindow.cpp	Tue Jun 15 06:08:48 2021 +0000
+++ b/widget/gtk/nsWindow.cpp	Tue Jun 15 06:10:21 2021 +0000
@@ -713,8 +713,6 @@
     mWaylandVsyncSource->Shutdown();
     mWaylandVsyncSource = nullptr;
   }
-
-  RemovePopupFromHierarchyList();
 #endif
 
   // It is safe to call DestroyeCompositor several times (here and
@@ -1339,15 +1337,8 @@
   if (aRemoveFromPopupList) {
     RemovePopupFromHierarchyList();
   }
-  if (aTemporaryHide) {
-    if (gtk_widget_is_visible(mShell)) {
-      mPopupTemporaryHidden = true;
-      HideWaylandWindow();
-    }
-  } else {
-    mPopupTemporaryHidden = false;
-    HideWaylandWindow();
-  }
+  mPopupTemporaryHidden = aTemporaryHide;
+  HideWaylandWindow();
 }
 
 void nsWindow::HideWaylandToplevelWindow() {
@@ -2083,12 +2074,7 @@
     return;
   }
 
-  // Only update setup for visible widgets - those are already tracked
-  // in layout hierarchy so we know their states from layout.
-  if (gtk_widget_is_visible(mShell)) {
-    MOZ_RELEASE_ASSERT(IsInPopupHierarchy());
-    UpdateWaylandPopupHierarchy();
-  }
+  UpdateWaylandPopupHierarchy();
 }
 
 void nsWindow::WaylandPopupMove(bool aUseMoveToRect) {
